### PR TITLE
Rewrite Worker

### PR DESCRIPTION
**Pull request by vogdb**
Depends on #88

### CODE REVIEW

**Comment by vogdb** [comment]
Pushing here just for backup. Some minor changes are still remaining.

**Comment by vogdb** [comment]
I haven't changed the logic itself. I've only split it among appropriate classes. The improvement of the logic itself will be done in a separate PRs.

**Comment by vogdb** [comment]
@tomaszdurka please review

**Comment by vogdb** [comment]
@tomaszdurka it's very short, almost one liner!

**Comment by tomaszdurka** [comment]
Should I just say looks good ? :)

**Comment by vogdb** [comment]
Well, we can start a long discussion here, if you want :)

Kidding. Just say lgtm!

File: lib/worker-connection.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+var _ = require('underscore');
+var validator = require('validator');
+var Subscriber = require('./subscriber');
+
+/**
+ * @param {SockJSConnection} connection
+ * @param {Worker} worker
+ * @constructor
+ */
+function WorkerConnection(connection, worker) {
```
Comment (tomaszdurka) [inline lib/worker-connection.js]: Just a bit confused by this class. I don't like so much when there is Worker with list of WorkerConnection and each of them has Worker dependency (circular dep 🐼 ).

You thought about avoiding that?

File: lib/channel.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * @param {String} id
+ * @constructor
+ */
+function Channel(id) {
+  this.id = id;
+  /** @type {Subscriber[]} */
+  this.subscribers = [];
+  /** @type {Message[]} */
+  this.msgs = [];
```
Comment (vogdb) [inline lib/channel.js]: I kept old names to make a transparent transition from the old code to the new.

File: lib/message.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+/**
+ * @param {Object} content
+ * @constructor
+ */
+function Message(content) {
+  /** @type {Object} */
+  this.content = content;
+  /** @type {Number} */
+  this.timestamp = new Date().getTime();
```
Comment (vogdb) [inline lib/message.js]: I kept old names to make a transparent transition from the old code to the new.

File: lib/worker-connection.js (hunk 1 above)
Comment (vogdb) [inline lib/worker-connection.js]: My thought is to make it as a separate step cause now the most necessary is to split one canvas of code on separate entities. The current implementation resembles the previous very close which helps to keep backward compatibility yet providing the necessary structuring.

**Comment by vogdb** [comment]
@tomaszdurka please review

**Comment by vogdb** [comment]
@tomaszdurka ?

**Comment by tomaszdurka** [comment]
On it now.

File: lib/worker-util.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * @param {Number} requestId
+ * @param {Object} channelsData
+ */
+WorkerUtil.sendUpStatusRequest = function(requestId, channelsData) {
+  process.send({type: 'up-status-request', data: {requestId: requestId, channels: channelsData}});
+};
+
+module.exports = WorkerUtil;
```
Comment (tomaszdurka) [inline lib/worker-util.js]: new line missing

File: lib/worker-util.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+function WorkerUtil() {
```
Comment (tomaszdurka) [inline lib/worker-util.js]: Let's make it a bit more clean in terms of oop. So rather have worker class and instance with process passed as depdenency - more testable and granular without global state.

So:
```js
function Worker(process) {
  this.process = process;
}
```

File: lib/channel-list.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
+var Channel = require('./channel');
+
+/**
+ * @constructor
+ */
+function ChannelList() {
```
Comment (tomaszdurka) [inline lib/channel-list.js]: Ok with this class, was fine before too 🐼 

If the class name is so explicit let's rename all methods to not contain channel keyword (as in comments below).

File: lib/channel-list.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  /** @type {Object} */
+  this._channels = {};
+}
+
+/**
+ * @param {String} channelId
+ * @returns {Boolean}
+ */
+ChannelList.prototype.hasChannel = function(channelId) {
```
Comment (tomaszdurka) [inline lib/channel-list.js]: -> has

File: lib/channel-list.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  return !!this._channels[channelId];
+};
+
+/**
+ * @param {String} channelId
+ * @returns {Channel}
+ */
+ChannelList.prototype.getChannel = function(channelId) {
```
Comment (tomaszdurka) [inline lib/channel-list.js]: -> get

File: lib/channel-list.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  return this._channels[channelId];
+};
+
+/**
+ * @returns {Object}
+ */
+ChannelList.prototype.getChannels = function() {
```
Comment (tomaszdurka) [inline lib/channel-list.js]: -> getAll

File: lib/channel-list.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  return this._channels;
+};
+
+/**
+ * @param {String} channelId
+ * @returns {Channel} channel
+ */
+ChannelList.prototype.createChannel = function(channelId) {
```
Comment (tomaszdurka) [inline lib/channel-list.js]: -> create

File: lib/channel-list.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  this._channels[channelId] = new Channel(channelId);
+  return this._channels[channelId];
+};
+
+/**
+ * @param {String} channelId
+ */
+ChannelList.prototype.delayedCloseChannel = function(channelId) {
```
Comment (tomaszdurka) [inline lib/channel-list.js]: -> delayedClose?

File: lib/channel.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  /** @type {Message[]} */
+  this._messages = [];
+  /** @type {Number|null} */
+  this.closeTimeout = null;
+}
+
+/**
+ * @param {WorkerConnection} connection
+ */
+Channel.prototype.unsubscribe = function(connection) {
```
Comment (tomaszdurka) [inline lib/channel.js]: Maybe let's be a bit more explicit here and name this `unsubscribeByConnection`?

File: lib/channel.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  this.subscribers = _.reject(this.subscribers, function(subscriber) {
+    return subscriber.connection === connection;
+  });
+};
+
+/**
+ * @returns {Boolean}
+ */
+Channel.prototype.hasSubscribers = function() {
```
Comment (tomaszdurka) [inline lib/channel.js]: As subscribers is array I would drop this method and simply use length outside, but as you wish here.

File: lib/worker-connection.js [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+  this.id = connection.id;
+
+  /** @type {SockJSConnection} */
+  this._connection = connection;
+
+  /** @type {ChannelList} */
+  this._channelList = channelList;
+
+  /** @type {String[]} */
+  this._сhannelIds = [];
```
Comment (tomaszdurka) [inline lib/worker-connection.js]: Do we need this list if we have channelList anyway?

File: lib/worker-connection.js [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+  this._startTimeout();
+}
+
+util.inherits(WorkerConnection, EventEmitter);
+
+/**
+ * @param {String} channelId
+ * @returns {Boolean}
+ */
+WorkerConnection.prototype.hasChannelId = function(channelId) {
```
Comment (tomaszdurka) [inline lib/worker-connection.js]: I would remove those methods and use `WorkerConnection.channels.has(channelId)` signature.

File: lib/worker-connection.js [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+ * @param {String} channelId
+ */
+WorkerConnection.prototype.addChannelId = function(channelId) {
+  this._сhannelIds.push(channelId);
+};
+
+/**
+ * @param {String} channelId
+ */
+WorkerConnection.prototype.removeChannelId = function(channelId) {
```
Comment (tomaszdurka) [inline lib/worker-connection.js]: Is this sufficient? Shouldn't we remove it from the channelList too? Same as above is channelList not enough?

File: lib/worker-connection.js (hunk 14 above)
Comment (tomaszdurka) [inline lib/worker-connection.js]: I looked into the usages below and seems that channelIds and channelList are always used together (unless missed something). Can we drop ids?

File: lib/channel.js (hunk 13 above)
Comment (vogdb) [inline lib/channel.js]: I will keep the method

File: lib/worker-connection.js (hunk 14 above)
Comment (vogdb) [inline lib/worker-connection.js]: I would keep the current logic. The change you want is very fundamental. `_сhannelIds` are channels of WorkerConnection. `_channelList` are channels of Worker. They might differ. `_channelList` is like a host to `_сhannelIds` of all its WorkerConnection's. We might to merge them but I would create a separate issue for that.

**Comment by vogdb** [comment]
@tomaszdurka please review

File: lib/worker.js [hunk 17]
```diff
@@ -1,328 +1,151 @@
...
+ * @param {Object}  [sslOptions]
+ * @constructor
+ */
+function Worker(port, sockjsClientUrl, sslOptions) {
+  /** @type {ChannelList} */
+  this._channelList = new ChannelList();
+  /** @type {Array} */
+  this._connections = [];
+  /** @type {WorkerSender} */
+  this._workerSender = new WorkerSender(process);
```
Comment (tomaszdurka) [inline lib/worker.js]: Last thing:
Let's put either WorkerSender or process one level up to the entrypoint?!

File: lib/channel-list.js (hunk 6 above)
Comment (tomaszdurka) [inline lib/channel-list.js]: https://github.com/cargomedia/socket-redis/blob/master/bin/socket-redis.js#L94

**Comment by vogdb** [comment]
@tomaszdurka please review

**Comment by tomaszdurka** [comment]
Great! Only 2 to go!

**Comment by vogdb** [comment]
Yes! One PR per weak!

**Comment by tomaszdurka** [comment]
I think it took 1.5...

**Comment by vogdb** [comment]
I mean the time that you have still ;-)

**Pull request merged**